cli/cmd/agent_aws-install_ec2ssm.go

```diff
@@ -0,0 +1,249 @@
+// License:: Apache License, Version 2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package cmd
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/service/ssm"
+	ssmtypes "github.com/aws/aws-sdk-go-v2/service/ssm/types"
+	"github.com/gammazero/workerpool"
+	"github.com/pkg/errors"
+	"github.com/spf13/cobra"
+)
+
+var (
+	agentInstallAWSSSMCmd = &cobra.Command{
+		Use:   "ec2ssm",
+		Args:  cobra.NoArgs,
+		Short: "Use SSM to securely install on EC2 instances",
+		RunE:  installAWSSSM,
+		Long: `This command installs the agent on all EC2 instances in an AWS account using SSM.
+
+To filter by one or more regions:
+
+    lacework agent aws-install ec2ssm --include_regions us-west-2,us-east-2
+
+To filter by instance tag:
+
+    lacework agent aws-install ec2ssm --tag TagName,TagValue
+
+To filter by instance tag key:
+
+    lacework agent aws-install ec2ssm --tag_key TagName
+
+To provide an agent access token of your choice, use the command 'lacework agent token list',
+select a token and pass it to the '--token' flag. This flag must be selected if the
+'--noninteractive' flag is set.
+
+    lacework agent aws-install ec2ssm --token <token>
+
+AWS credentials are read from the following environment variables:
+- AWS_ACCESS_KEY_ID
+- AWS_SECRET_ACCESS_KEY
+- AWS_SESSION_TOKEN (optional)
+- AWS_REGION (optional)`,
+	}
+)
+
+func init() {
+	// 'agent aws-install ec2ssm' flags
+	agentInstallAWSSSMCmd.Flags().StringVar(&agentCmdState.InstallTagKey,
+		"tag_key", "", "only install agents on infra with this tag key set",
+	)
+	agentInstallAWSSSMCmd.Flags().StringSliceVar(&agentCmdState.InstallTag,
+		"tag", []string{}, "only install agents on infra with this tag",
+	)
+	agentInstallAWSSSMCmd.Flags().StringSliceVarP(&agentCmdState.InstallIncludeRegions,
+		"include_regions", "r", []string{}, "list of regions to filter on",
+	)
+	agentInstallAWSSSMCmd.Flags().StringVar(&agentCmdState.InstallAgentToken,
+		"token", "", "agent access token",
+	)
+	agentInstallAWSSSMCmd.Flags().IntVarP(
+		&agentCmdState.InstallMaxParallelism,
+		"max_parallelism",
+		"n",
+		50,
+		"maximum number of workers executing AWS API calls, set if rate limits are lower or higher than normal",
+	)
+	agentInstallAWSSSMCmd.Flags().StringVar(
+		&agentCmdState.InstallBYORole,
+		"iam_role_name",
+		"",
+		"IAM role name (not ARN) with SSM policy, if not provided then an ephemeral role will be created",
+	)
```

Comment:
One other option -- add a flag that just skips installing or verifying and SSM instance profile. If customers are using instance profiles then it's possible that they already have enabled SSM access.

Documenting that they can either have the CLI create the role/profile _or_ provide it themselves seems like it would bump customer use cases to maybe 90% or 95%. The big thing left out would be customers who require that SSM only be provided with specific Command documents.

Reply:
Yeah, that makes sense, basically saying that if the customer has everything setup for SSM already, then give them the option. It's basically what the logic is doing by checking if the existing infra works.

Reply:
927f83530ea0052ed3aa8519fb41afb909494444